### ISSUE TITLE
Add recycle bin functionality

### ISSUE THREAD

**Issue by rfrenchseti** [SETI/rms-opus#845]
Due to issue #739, we have currently disabled the ability to remove items from the cart while on the cart tab. We want to replace this with a new "recycle bin" functionality:

Back end changes:

- DONE: Modify the "cart" table to include a boolean "recycled" column. Thus an observation can be not in the cart, in the cart but not in the recycle bin, or in the cart and in the recycle bin.
- DONE: Modify the "dataimages.json" API to return the tri-state status of the observation by renaming "in_cart" to "cart_state" and making the return values False, "cart", and "recycle"
- DONE: Modify the cart remove/removerange API calls to take a new "recyclebin=0/1" argument. If 0, the behavior is as it is now. If 1, remove takes an obs in the cart and sets the recycled field to true; add always takes an obs where the recycled field is true and sets it to false.
- DONE: Modify the cart status API call to return both the number of items in the cart and the number of items in the recycle bin.
- DONE: Modify the cart reset API call to take the recyclebin=0/1 argument to say whether to empty the entire cart (including the recycle bin) or just the recycle bin.
- DONE: Modify addall.json to move items from the recycle bin back to the main cart.

Front end changes:

- Nothing has to change on the browse gallery/table except items in "recycle" state should not be highlighted.
- On the cart tab, items in "recycle" state should have a large "recycle bin" icon plastered across them, just like we do with the binoculars.
- On the cart tab, the "add to cart" icon should be "move to recycle bin" and have a different icon; likewise for "remove from cart" being "restore from recycle bin". Same for the hamburger menu items and the range options.
- On the cart tab, ctrl+click should toggle recycle bin state.
- On the cart tab, there need to be new buttons "Empty Recycle Bin" (which causes a hard refresh) and "Restore Recycle Bin". Somewhere related to these buttons there should be an indication of how many items are in the recycle bin.
- The cart status badge needs to display only the number of items in the cart, not the number of items in the recycle bin.

For responsiveness, the buttons that say "Empty Recycle Bin" etc might be combined together into some kind of drop-down menu. Also as the screen gets narrower the words might go away and be replaced by only icons.

**Comment by rfrenchseti** [comment]
Log of backend changes:

- The "cart" table has a new "recycled" column. Before using these changes, you must regenerate the "cart" table by doing `python main_opus_import.py --create-cart`
- Any observation may be in one of three states: Not in the cart at all, in the cart but not in the recycle bin (recycled=0), or in the recycle bin (recycled=1). An observation can never be in two places at once.
- `__cart/add.json` and `__cart/remove.json` can now take a single opus_id (`?opusid=<x>`) or a list of opuis_ids (`?opusid=<x>,<y>,<z>`).
- `__cart/remove.json` and `removerange.json` now take an optional `?recyclebin=0/1` argument. If omitted it defaults to 0. The basic idea is that add always puts something in the cart outside of the recycle bin, and remove either moves something back from the recycle bin, or deletes it entirely.
- `__cart/addall.json` can work with `view=browse` (in which case it uses the search results) or `view=cart` (in which case it uses the contents of the cart). In the latter case, there is an optional `?recyclebin=0/1` that says whether to use just the main cart, or the cart+recycle bin. Doing `__cart/addall.json?view=cart&recyclebin=0` is silly and does nothing. Doing `__cart/addall.json?view=cart&recyclebin=1` is the way to move all recycled items back to the main cart.
- The state transitions for add/addrange/remove/removerange/addall are:

```
    add/addrange (recyclebin option ignored):
        Not previously in cart                  Add to cart recycled=0
        Previously in cart recycled=0           No effect
        Previously in cart recycled=1           Set recycled=0
        Bad opus_id                             Error

    addall recyclebin=0 (this means to take "all" from browse results or cart
                         depending on "browse=")
        Not previously in cart                  Add to cart recycled=0
        Previously in cart recycled=0           No effect
        Previously in cart recycled=1           Set recycled=0

    addall recyclebin=1 (this means to take "all" from browse results or
                         cart+recycle bin depending on "browse=")
        Not previously in cart                  Add to cart recycled=0
        Previously in cart recycled=0           No effect
        Previously in cart recycled=1           Set recycled=0

    remove/removerange recyclebin=0
        Not in cart                             No effect
        In cart recycled=0                      Remove from cart
        In cart recycled=1                      Remove from cart
        Bad opus_id                             No effect

    remove/removerange recyclebin=1
        Not in cart                             No effect
        In cart recycled=0                      Set recycled=1
        In cart recycled=1                      No effect
        Bad opus_id                             Error
```
- `__cart/status.json` always returns both `{'count': X}`, which is the number of obs in the cart but NOT in the recycle bin, and `{'recycled_count': X}`, which is the number of obs in the recycle bin. This is also true for add/remove/etc.
- `dataimages.json` replaces the `in_cart` flag with a new `cart_state` flag, which can be `false`, `"cart"`, or `"recycle"`. The same is true for `images.json`.
- `__cart/reset.json` now takes an optional `?recyclebin=0/1` argument. If omitted it defaults to 0. If `?recyclebin=1`, then only observations in the recycle bin are removed; otherwise all observations (normal cart and recycle bin) are removed. It also no longer resets the Django session_id.

**Comment by rfrenchseti** [comment]
Example uses of the new API variants:

```
    def test__api_cart_add_remove_recyclebin(self):
        "[test_cart_api.py] /__cart/add&remove: recyclebin"
        url = '/opus/__cart/reset.json'
        self._run_status_equal(url, 200)
        url = '/opus/__cart/addall.json?volumeid=COVIMS_0006&reqno=456'
        expected = {'recycled_count': 0, 'count': 3544, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?volumeid=COVIMS_0006&range=co-vims-v1484504505_vis,co-vims-v1484506475_vis&reqno=456'
        expected = {'recycled_count': 0, 'count': 3539, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?view=cart&range=co-vims-v1484506705_vis,co-vims-v1484507800_ir&reqno=456&recyclebin=0'
        expected = {'recycled_count': 0, 'count': 3535, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?view=cart&range=co-vims-v1484509868_ir,co-vims-v1484510890_vis&reqno=456&recyclebin=1'
        expected = {'recycled_count': 8, 'count': 3527, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/addrange.json?view=cart&range=co-vims-v1484509868_vis,co-vims-v1484510044_ir&reqno=456'
        expected = {'recycled_count': 6, 'count': 3529, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/addrange.json?view=browse&range=co-vims-v1484510714_ir,co-vims-v1484510890_ir&reqno=456'
        expected = {'recycled_count': 3, 'count': 3532, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/status.json?reqno=456'
        expected = {'recycled_count': 3, 'count': 3532, 'reqno': 456}
        self._run_json_equal(url, expected)

    def test__api_cart_recyclebin_reset(self):
        "[test_cart_api.py] /__cart/reset: recyclebin"
        url = '/opus/__cart/reset.json'
        self._run_status_equal(url, 200)
        url = '/opus/__cart/addall.json?volumeid=COVIMS_0006&reqno=456'
        expected = {'recycled_count': 0, 'count': 3544, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?view=cart&range=co-vims-v1484509868_ir,co-vims-v1484510890_vis&reqno=456&recyclebin=1'
        expected = {'recycled_count': 8, 'count': 3536, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/reset.json?recyclebin=1'
        self._run_status_equal(url, 200)
        url = '/opus/__cart/status.json?reqno=456'
        expected = {'recycled_count': 0, 'count': 3536, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/reset.json?recyclebin=0'
        self._run_status_equal(url, 200)
        url = '/opus/__cart/status.json?reqno=456'
        expected = {'recycled_count': 0, 'count': 0, 'reqno': 456}
        self._run_json_equal(url, expected)

    def test__api_cart_addall_recyclebin(self):
        "[test_cart_api.py] /__cart/addall: recyclebin"
        url = '/opus/__cart/reset.json'
        self._run_status_equal(url, 200)
        url = '/opus/__cart/addall.json?volumeid=COVIMS_0006&reqno=456'
        expected = {'recycled_count': 0, 'count': 3544, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?volumeid=COVIMS_0006&range=co-vims-v1484504505_vis,co-vims-v1484506475_vis&reqno=456'
        expected = {'recycled_count': 0, 'count': 3539, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/removerange.json?view=cart&range=co-vims-v1484509868_ir,co-vims-v1484510890_vis&reqno=456&recyclebin=1'
        expected = {'recycled_count': 8, 'count': 3531, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/addall.json?view=cart&reqno=456' # Does nothing
        expected = {'recycled_count': 8, 'count': 3531, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/addall.json?view=cart&reqno=456&recyclebin=1'
        expected = {'recycled_count': 0, 'count': 3539, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/addall.json?view=cart&reqno=456&recyclebin=1'
        expected = {'recycled_count': 0, 'count': 3539, 'error': False, 'reqno': 456}
        self._run_json_equal(url, expected)
        url = '/opus/__cart/status.json?reqno=456'
        expected = {'recycled_count': 0, 'count': 3539, 'reqno': 456}
        self._run_json_equal(url, expected)
```

**Comment by rfrenchseti** [comment]
Current issues:

- On the cart tab, the trashcan icons should be recycle bins
- On the hamburger menu, "start move range to recycle bin" should end with "here" like the normal start range item does
- The little pop-up messages at the bottom for start/end range are wrong for move to/restore from recycle bin
- I don't think we want the recycle bin icons to show up on the browse results tab, just the cart tab
- The recycle bin icons on the browse results tab get confused anyway - if you add a range to the cart that includes recycled items, you end up with thumbnails that are both highlighted AND have a recycle bin icon
- Same for the table view on the browse results tab - right now the recycled items are green, but not lined through. I don't think they should be green at all.
- On the cart tab, table view, there is a green line sort of vaguely between the checkbox and the hamburger, which doesn't look good.
- On the cart tab, table view, I think there should be some way to indicate this is a recycle bin item, like a small icon over the checkbox or a line through the checkbox or something similar.
- For the confirm restore recycle bin modal, I'd flip the last two clauses...restore all observations from the recycle bin to your cart
- Restore recycle bin doesn't update the table view properly
- Restore recycle bin doesn't entirely update the gallery view properly - it removes the recycle icons but the state is still wrong, so the menu options do the wrong thing
- The Getting Started Guide should be updated.
- I think some kind of formatting needs to be done on the Total Size/Total Files line. Maybe center over the buttons. Maybe add a small amount of vertical whitespace between the text and the buttons.
- There needs to be some kind of responsiveness fix for the Download pane.
- When the Download pane is collapsed, the "Download Options" button needs to be relabeled so it's obvious you have to go there to download at all. Maybe something really obvious like "Download Here"
- When the Download pane is collapsed, the headings get changed since now the main pane heading is "Download Options" and now there's no separate "Download" or "Download Options" headers.
- Empty/restore recycle big and empty cart should be sensitive to whether they are actually useful and deactivated otherwise
- Add shadow around recycle icon
- Replace trashcan icon and cart icon with recycle and restore icons
- Add icons back to bottom of slideshow
- Fix #883 
- Clarify nav text for restore and empty of recycle bin; rename restore recycle bin to "restore from recycle bin"

**Comment by rfrenchseti** [comment]
- Also add shadow around binoculars while you're at it
- Fix wraparound of product type names so that they always stay to the right of the (i)
